### DISCUSSION TITLE
Possible to fully remove Pino?

### DISCUSSION THREAD

**Discussion by prescience-data** [fastify/fastify#2996] [Q&A]
I was wondering if there is a way to fully remove Pino from the library as it is the only dependency that is breaking my ESBuild bundling process.

This means I need to exclude it from bundling, which in turn means when I install the bundled package in my container, it needs to pull Pino and all of its dependencies making the node_modules directory substantially larger than it would otherwise be. 

I've disabled Pino with:

```typescript
export const server: FastifyInstance = fastify({
  logger: false
})
```

However, there is something fundamentally bound within Fastify which is accessed during the build / tree-shaking process.

Possibly a long shot, but wanted to put it out there in case anyone else has managed to achieve this.

**Comment by zekth** [comment]
Technically speaking it would be doable but would require some changes in here: https://github.com/fastify/fastify/blob/0ae362b7d81b0e823fa7c9c87fc6d4655569d822/fastify.js

But the question is why is it breaking your bundling process? May you share the issue? I don't think it would be really valuable to allow people to not use pino.

**Reply by prescience-data** [reply]
Regarding the build process:

### ESBuild config:
```js
require("esbuild").buildSync({
  bundle: true,
  minify: true,
  format: "cjs",
  keepNames: true,
  sourcemap: true,
  platform: "node",
  logLevel: "error",
  target: ["node14"],
  external: ["fsevents", "bullmq", "pino"],
  outfile: "dist/run.js",
  entryPoints: ["src/index.ts"]
})
```
When `pino` is excluded, the bundle launches successfully and the server starts.

When it is removed from the `external` exclusions (ie attempts to bundle it), the bundle throws the (unhelpful) error:
```
ReferenceError: e is not defined
```

Generally, I've found this is due to a circular dependency within the offending package somewhere.

**Reply by prescience-data** [reply]
Additionally, I don't mean to actually remove it - just have it be properly abstracted (so if a user passes `false` or their own logger to the `logger` launch option, it never actually touches the `pino` package when tree-shaking.

It seems like there is a tight coupling within Fastify that implements Pino even when the user has requested no logger be instantiated.

**Reply by Eomm** [reply]
Did you try to pass an empty custom logger?

https://stackoverflow.com/a/55266062/3309466

**Comment by mcollina** [comment]
Fastify aims to ship a logger out of the box and a few "old" members of this organization are the ones you are the ones maintaining pino as well.

We believe not shipping a logger provide a worse experience than shipping one.

FWIW, if there is a bug somewhere in pino, feel free to send a PR.

**Reply by prescience-data** [reply]
Hey! Sorry, for clarity I didn't mean to actually remove it - just to only load it if no other logger is provided. (ie allow it to be excluded when tree-shaking)

Adding package weight for a full logging library that is not used by the application (if injecting custom logger with `logger` setting) seems unnecessary, but I could easily be misunderstanding how deeply integrated Pino is. 

I had figured that if a custom `logger` was passed in, Pino would be effectively neutralised / untouched?

**Reply by mcollina** [reply]
Yes exactly.

I'm not sure what you would like to achieve is possible, as conditional loading is not supported by bundlers as well.

**Reply by prescience-data** [reply]
I spent a bit of time trying to figure out where the issue is arising during modification (to start a Pino PR), but because all the variables get mangled it's super hard to figure out where exactly it throws. 
In any case, as mentioned below, I managed to solve it by disabling minification so that is a win!

**Comment by jsumners** [comment]
Is "bundling" actually necessary?

**Reply by prescience-data** [reply]
Normally no, in this instance where container size is critical and there are several non-published dependency packages (monorepo), yes.
fwiw I managed to get it to avoid throwing error by disabling minification, which although comes at a cost, is better than pulling in all the additional dependencies. 👍
BullMQ is the only remaining problem package now, and that fails when bundling even with the minifier disabled. 😔
